delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.delta
+
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.delta.RapidsRepartitionByFilePath.FILE_PATH_FIELD
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.{PartitionCoalescer, PartitionGroup, RDD}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, UnaryNode}
+import org.apache.spark.sql.catalyst.plans.physical.{HashPartitioning, Partitioning}
+import org.apache.spark.sql.execution.{SparkPlan, SparkStrategy, UnaryExecNode}
+import org.apache.spark.sql.types.{LongType, StringType, StructField}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object RapidsRepartitionByFilePath {
+  val ROW_ID_COL = "_metadata_row_id"
+  val ROW_ID_FIELD = StructField(ROW_ID_COL, LongType, nullable = false)
+  val FILE_PATH_COL = "_metadata_file_path"
+  val FILE_PATH_FIELD = StructField(FILE_PATH_COL, StringType, nullable = false)
+}
+
+case class RapidsRepartitionByFilePath(optNumPartition: Option[Int],
+    child: LogicalPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+
+  override protected def withNewChildInternal(newChild: LogicalPlan): LogicalPlan =
+    copy(child = newChild)
+}
+
+case class RapidsRepartitionByFilePathExec(optNumPartition: Option[Int],
+    child: SparkPlan) extends UnaryExecNode {
+
+  private lazy val numPartitions = optNumPartition.getOrElse(conf.numShufflePartitions)
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    throw new UnsupportedOperationException("RapidsParquetFileScanExec should not be executed")
+  }
+
+  override def output: Seq[Attribute] = child.output
+
+  override protected def withNewChildInternal(newChild: SparkPlan): SparkPlan =
+    copy(child = newChild)
+
+  override def outputPartitioning: Partitioning = {
+    child.output.find(_.name == FILE_PATH_FIELD.name) match {
+      case Some(attr) => HashPartitioning(Seq(attr), numPartitions)
+      case None => throw new IllegalArgumentException(
+        s"Cannot find ${FILE_PATH_FIELD.name} in the output of the child plan")
+    }
+  }
+}
+
+case class GpuRapidsRepartitionByFilePathExec(optNumPartition: Option[Int],
+    gpuPartitionIdExpr: GpuExpression,
+    child: SparkPlan) extends UnaryExecNode with GpuExec {
+  private lazy val numPartitions = optNumPartition.getOrElse(conf.numShufflePartitions)
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = {
+    child.output.find(_.name == FILE_PATH_FIELD.name) match {
+      case Some(attr) => HashPartitioning(Seq(attr), numPartitions)
```

Comment:
So this new node is likely going to inject an extra shuffle because of this partitioning.  Intentional?